src/components/NewModal/index.js

```diff
@@ -0,0 +1,168 @@
+
+import { ModalProvider } from "./context";
+
+export default class FilterModal extends React.Component {
+  static propTypes = {
+    actionBar: PropTypes.node,
+    bottomActionBar: PropTypes.node,
+    children: PropTypes.node.isRequired,
+    isOpened: PropTypes.bool,
+    onRequestClose: PropTypes.func,
+    onScroll: PropTypes.func
+  };
+
+  static defaultProps = {
+    actionBar: <DefaultActionBar />,
+    bottomActionBar: null,
+    isOpened: true,
+    onRequestClose: null,
+    onScroll: null
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    if (props.isOpened !== state.isOpened) {
+      return {
+        isOpened: props.isOpened
+      };
+    }
+
+    return null;
+  }
+
+  state = {
+    isOpened: true,
+    actionBarShadow: false,
+    bottomActionBarShadow: false
+  };
+
+  componentDidMount() {
+    const actionBar = this.actionBarRef.current;
+    const bottomActionBar = this.bottomActionBarRef.current;
+    const content = this.contentRef.current;
+
+    if (actionBar) {
+      content.style.marginTop = actionBar.clientHeight;
+    }
+
+    if (bottomActionBar) {
+      content.style.marginBottom = bottomActionBar.clientHeight;
+    }
+
+    this.updateShadows();
+  }
+
+  actionBarRef = React.createRef();
+  containerRef = React.createRef();
+  contentRef = React.createRef();
+  bottomActionBarRef = React.createRef();
+
+  openModal = () => this.setState({ isOpened: true });
+
+  closeModal = async () => {
+    const { onRequestClose } = this.props;
+
+    const requestApproved = onRequestClose && (await onRequestClose());
+
+    if (!onRequestClose || requestApproved) {
+      this.setState({ isOpened: false });
+    }
+  };
+
+  handleScroll = e => {
+    const { onScroll } = this.props;
+
+    if (onScroll) {
+      onScroll(e);
+    }
+
+    this.updateShadows();
+  };
+
+  updateShadows = () => {
+    this.updateActionBarShadow();
+
+    const bottomActionBar = this.bottomActionBarRef.current;
+
+    if (bottomActionBar) {
+      this.updateBottomActionBarShadow();
+    }
+  };
+
+  updateActionBarShadow = () => {
+    const container = this.containerRef.current;
+
+    const actionBarShadow = Boolean(container.scrollTop);
+
+    if (actionBarShadow !== this.state.actionBarShadow) {
+      this.setState({
+        actionBarShadow
+      });
+    }
+  };
+
+  updateBottomActionBarShadow = () => {
+    const actionBar = this.actionBarRef.current;
+    const container = this.containerRef.current;
+    const content = this.contentRef.current;
+
+    const bottomActionBar = this.bottomActionBarRef.current;
+
+    const bottomActionBarShadow = Boolean(
+      content.clientHeight - container.scrollTop >
+        container.clientHeight -
+          bottomActionBar.clientHeight -
+          actionBar.clientHeight
+    );
+
+    if (bottomActionBarShadow !== this.state.bottomActionBarShadow) {
+      this.setState({
+        bottomActionBarShadow
+      });
+    }
+  };
+
+  render() {
```

Comment:
I think we should rely on composition and pass the `ActionBar` and `BottomActionBar` in as `children`, if desired. This should reduce the need for `refs`, restyling, and checking element height. Instead, we can create an example in the catalog that showcases how to utilize this `Modal` with `ActionBar` and `BottomActionBar` passed in as `children`.

Reply:
It was done this way because (currently) we have only modals that have top/bottom bars, scrollable body, and shadows, so this one is a really common case for us.

We have two options:
1) Omit pass `null` to actionBar and bottomActionBar, so modal will not draw and calculate shadows
2) Create `ScrollableModalWithActionBarShadowsModal` component, and move shadow-related logic into it.

As for me, I would like to stick with option #1, rather having almost-the-same modal components, even if they use composition to increase code reusage.

Reply:
Let's proceed with this implementation given our time constraints.